### PR TITLE
Add dataminimisation for payments, sales and pizzas

### LINKED ISSUE #880: Data minimization on payments and event registrations after 7 years.
In GitLab by @JobDoesburg on Jun 2, 2019, 15:35

### One-sentence description

Data minimization on payments and event registrations after 7 years.

### Motivation

The privacy office of the RU has indicated that there are no grounds for Thalia to keep data of payments or event participation after 7 years (the legal obligatory time to keep this information). So in order to be fully compliant to the AVG/GDPR we should maybe delete this information after 7 years.

### Desired functionality

7 years after an event, information of who attended what activity should be removed and whether they paid, and how. 
It would be nice if we can keep track of the amount of participants at that activity.

### Suggested implementation

Apart from that I can imagine this is not quite easy, I have no idea.

Note that this request is only based on information provided by the Radboud University privacy office. But the Technicie I'm sure also have an grounded opinion about this. So if you guys think it is unnecessary to remove this data because of reasons, and you think it is compliant to the AVG/GDPR that is fine as well! This is just passing on information actually.

### LINKED ISSUE #1653: Minimise sales order data
### Is your feature request related to a problem? Please describe.
The new `sales` app stores quite some member credentials that we don't need to keep for a long time

### Describe the solution you'd like
Anonymise the orders (setting the Member field to NULL) after a while (30 days? 1 year? directly when a shift closes)

Note that this also needs an update in the privacy statement!

### Motivation
Data minimization 

### Describe alternatives you've considered
Right now this will happen when a member object is deleted, but that takes too long in my opinion. There's no need to keep an order history in my opinion.

### Additional context
- We shouldn't delete the orders themselves, we want to keep them for multiple reasons
- This is no true data minimization, as orders will have payments, and those payments still have a user connected, but that's another issue.

### PR DESCRIPTION
**Pull request by JobDoesburg**
Closes #880 and closes #1653

### Summary
Adds data minimization for payments, sales and pizzas

### How to test
1. Have old data
2. Execute dataminimisation
3. Verify the data is minimized 
4. Verify that things don't crash after minimization

### CODE REVIEW

**Comment by JobDoesburg** [comment]
@pingiun do you know what's wrong with postgres

**Comment by pingiun** [comment]
> @pingiun do you know what's wrong with postgres

```
======================================================================
243
ERROR: test_data_minimisation (payments.tests.test_services.ServicesTest) [Payments that are 7 years old must be minimised]
244
----------------------------------------------------------------------
245
Traceback (most recent call last):
246
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
247
    return self.cursor.execute(sql, params)
248
psycopg2.errors.NumericValueOutOfRange: numeric field overflow
249
DETAIL:  A field with precision 4, scale 2 must round to an absolute value less than 10^2.
```

**Comment by JobDoesburg** [comment]
> > @pingiun do you know what's wrong with postgres
> 
> ```
> ======================================================================
> 243
> ERROR: test_data_minimisation (payments.tests.test_services.ServicesTest) [Payments that are 7 years old must be minimised]
> 244
> ----------------------------------------------------------------------
> 245
> Traceback (most recent call last):
> 246
>   File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
> 247
>     return self.cursor.execute(sql, params)
> 248
> psycopg2.errors.NumericValueOutOfRange: numeric field overflow
> 249
> DETAIL:  A field with precision 4, scale 2 must round to an absolute value less than 10^2.
> ```

Yeah I see that, but it doesn't happen locally and I don't see on what field it's actually crashing... It's not the `amount`...

**Comment by pingiun** [comment]
> > > @pingiun do you know what's wrong with postgres
> > 
> > 
> > ```
> > ======================================================================
> > 243
> > ERROR: test_data_minimisation (payments.tests.test_services.ServicesTest) [Payments that are 7 years old must be minimised]
> > 244
> > ----------------------------------------------------------------------
> > 245
> > Traceback (most recent call last):
> > 246
> >   File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
> > 247
> >     return self.cursor.execute(sql, params)
> > 248
> > psycopg2.errors.NumericValueOutOfRange: numeric field overflow
> > 249
> > DETAIL:  A field with precision 4, scale 2 must round to an absolute value less than 10^2.
> > ```
> 
> Yeah I see that, but it doesn't happen locally and I don't see on what field it's actually crashing... It's not the `amount`...

It doesn't happen locally because sqlite stores rationals as strings I believe

**Comment by pingiun** [comment]
A field with precision 4 scale 2 can only store 4 significant digits i believe, so max 99.99

**Comment by JobDoesburg** [comment]
> A field with precision 4 scale 2 can only store 4 significant digits i believe, so max 99.99

The weird thing is that  we should have 6 digits and 2 decimals...

File: website/payments/services.py [hunk 1]
```diff
@@ -189,3 +189,16 @@ def send_tpay_batch_processing_emails(batch):
             },
         )
     return len(member_payments)
+
+
+def execute_data_minimization(dry_run=False):
```
Comment (se-bastiaan) [inline website/payments/services.py:194]: minimisation or minimization?

I think the member code use the `s`.

File: website/pizzas/services.py [hunk 2]
```diff
@@ -29,3 +31,22 @@ def can_change_order(member, food_event):
...
+    queryset_members = FoodOrder.objects.filter(created_at__lte=deletion_period).filter(
+        payer__isnull=False
+    )
+    queryset_nonmembers = (
+        FoodOrder.objects.filter(created_at__lte=deletion_period)
+        .filter(name__isnull=False)
+        .exclude(name="<deleted>")
+    )
+    if not dry_run:
+        queryset_members.update(payer=None)
```
Comment (se-bastiaan) [inline website/pizzas/services.py:50]: `payer` does not exist on this model

File: website/pizzas/services.py [hunk 3]
```diff
@@ -29,3 +31,22 @@ def can_change_order(member, food_event):
...
+        queryset_nonmembers.update(name="<deleted>")
+    return queryset_members.union(queryset_nonmembers).all()
```
Comment (se-bastiaan) [inline website/pizzas/services.py:52]:
```suggestion
    queryset = FoodOrder.objects.filter(created_at__lte=deletion_period).exclude(name="<removed>")
    if not dry_run:
        queryset.update(member=None, name="<removed>")
    return queryset
```

Don't see why you need two queries to do this.

File: website/pizzas/services.py [hunk 4]
```diff
@@ -29,3 +31,22 @@ def can_change_order(member, food_event):
...
         and is_organiser(member, food_event.event)
     )
+
+
+def execute_data_minimization(dry_run=False):
+    """Anonymizes pizzas orders older than 3 years."""
+    # Sometimes years are 366 days of course, but better delete 1 or 2 days early than late
+    deletion_period = timezone.now().date() - timezone.timedelta(days=(365 * 3))
+
+    queryset_members = FoodOrder.objects.filter(created_at__lte=deletion_period).filter(
```
Comment (se-bastiaan) [inline website/pizzas/services.py:41]: Food orders don't have a `created_at`.

File: website/pizzas/services.py [hunk 5]
```diff
@@ -29,3 +31,22 @@ def can_change_order(member, food_event):
         and member.has_perm("pizzas.change_foodorder")
         and is_organiser(member, food_event.event)
     )
+
+
+def execute_data_minimization(dry_run=False):
+    """Anonymizes pizzas orders older than 3 years."""
+    # Sometimes years are 366 days of course, but better delete 1 or 2 days early than late
+    deletion_period = timezone.now().date() - timezone.timedelta(days=(365 * 3))
```
Comment (se-bastiaan) [inline website/pizzas/services.py:39]: Maybe this should be a setting or global thing somewhere? Would that make sense?

File: website/sales/services.py [hunk 6]
```diff
@@ -19,3 +21,16 @@ def is_manager(member, shift):
...
     return False
+
+
+def execute_data_minimization(dry_run=False):
+    """Anonymizes orders older than 3 years."""
+    # Sometimes years are 366 days of course, but better delete 1 or 2 days early than late
+    deletion_period = timezone.now().date() - timezone.timedelta(days=(365 * 3))
+
+    queryset = Order.objects.filter(created_at__lte=deletion_period).exclude(
+        member__isnull=True
```
Comment (se-bastiaan) [inline website/sales/services.py:32]:
```suggestion
        payer__isnull=True
```

File: website/payments/services.py [hunk 7]
```diff
@@ -189,3 +189,16 @@ def send_tpay_batch_processing_emails(batch):
...
+    """Anonymizes payments older than 7 years."""
+    # Sometimes years are 366 days of course, but better delete 1 or 2 days early than late
+    deletion_period = timezone.now().date() - timezone.timedelta(days=(365 * 7))
+
+    queryset = Payment.objects.filter(created_at__lte=deletion_period).exclude(
+        paid_by__isnull=True
+    )
+    if not dry_run:
+        queryset.update(paid_by=None)
```
Comment (se-bastiaan) [inline website/payments/services.py:203]:
```suggestion
        queryset.update(paid_by=None, processed_by=None)
```

This is also personal information.

File: website/payments/services.py [hunk 8]
```diff
@@ -189,3 +189,16 @@ def send_tpay_batch_processing_emails(batch):
...
+    return queryset.all()
```
Comment (se-bastiaan) [inline website/payments/services.py:204]:
```suggestion
    return queryset
```

don't think you need that

**Review by se-bastiaan — CHANGES_REQUESTED** [review]
🚒

**Comment by JobDoesburg** [comment]
Another thing I just thought about: Django Admin LogEntry's might need to be anonymized too

File: website/pizzas/services.py (hunk 5 above)
Comment (JobDoesburg) [reply website/pizzas/services.py:39]: Let me create a new issue for this, because I like the idea!

Comment (JobDoesburg) [reply website/pizzas/services.py:39]: See #2099

File: website/payments/templates/payments/payment_list.html [hunk 9]
```diff
@@ -92,6 +92,9 @@ <h5 class="mb-4">
                 {% trans "You have not made any payments during this period." as no_payments %}
                 {% alert 'info' no_payments %}
             {% endif %}
+            <span class="mt-3 font-italic">
+                {% trans "Thalia only stores your payments for 7 years." %}
```
Comment (se-bastiaan) [inline website/payments/templates/payments/payment_list.html:96]: Maybe add that this is a legal requirement?

**Review by se-bastiaan — APPROVED** [review]

File: website/payments/templates/payments/payment_list.html (hunk 9 above)
Comment (JobDoesburg) [reply website/payments/templates/payments/payment_list.html:96]: Meh, they can read it in the privacy policy. The reason I added this here was just to explain why you don't see everything.

**Review by pingiun — APPROVED** [review]

**Pull request merged**